fix(platform-xmpp): Attempt to properly determine actor type incoming updates

silverbucket opened the pull request:
## Problem

Fixes #679. Presence updates from XMPP MUC rooms were always typed as `actor.type: "person"` instead of `"room"`. A JID like `kosmos@kosmos.chat` has no pattern that identifies it as a conference room, so heuristics based on the JID string will always fail for real-world deployments.

The previous attempt used service discovery (disco#info) to determine actor type reactively at presence-receive time. This was fragile: transient timeouts permanently cached the wrong type, IQ error responses fell through to `"person"` silently, and any JID without `conference.` / `muc.` / `chat.` in the domain was misclassified.

## Solution

Track joined rooms proactively. `join()` already knows the room JID — register it. `presence()` checks that set synchronously.

- `join()` adds the room bare JID to `this.__knownRooms`
- `leave()` removes it
- `presence()` checks `__knownRooms` — if present, `actor.type = "room"`, otherwise `"person"`
- Removed entirely: `determineActorType()`, `jidTypeCache`, `DISCOVERY_TIMEOUT`, `Math.random()` ID generation, `sendReceive` call, heuristic fallback, `async` on `presence()`

## Why this works

XMPP only delivers presence updates for JIDs you are in a relationship with — rooms you have joined, or roster contacts. There is no realistic case where an unknown room sends an unsolicited presence update. So the join/leave lifecycle is the correct and only reliable source of truth for whether a JID is a room.

## Testing

Unit tests cover:
- Person typing for unknown JIDs
- Room typing after `join()` registration
- Resource stripping before lookup (`room@server/nick` → `room@server`)
- Revert to person after `leave()` deregistration
- Multiple rooms tracked independently
- `join()` and `leave()` update `__knownRooms` correctly

Integration test: after joining `test@prosody`, waits for a presence update from that JID and asserts `actor.type === "room"`.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Presence now classifies actors as "room" or "person" so updates reflect room vs. individual sources.

* **Bug Fixes**
  * Presence debug messages include the resolved actor type.
  * Join/leave now track room membership so unavailable presence is addressed and cleared correctly.

* **Tests**
  * Added unit and integration tests covering room presence typing, join/leave membership tracking, and related presence behaviors.

<!-- review_stack_entry_start -->

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/sockethub/sockethub/pull/915?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<!-- review_stack_entry_end -->
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

silverbucket committed:
try to properly determine actor type on some incoming room presence updates

silverbucket committed:
adjustments based on copilot review

silverbucket committed:
set constant

silverbucket committed:
update tests based on feedback

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

raucao commented:
Still getting "person" type for the room actor:

<img width="1156" height="407" alt="Screenshot From 2026-01-08 17-31-59" src="https://github.com/user-attachments/assets/5fd12d49-8e77-4a69-a1f7-1443898a9c3d" />

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

raucao reviewed (changes requested):
It's still the same:

<img width="1103" height="237" alt="Image" src="https://github.com/user-attachments/assets/a2fece1d-1dca-4c91-b9ba-327572fcd5b5" />
  
You can reproduce and debug it directly in Hyperchannel using the Ember upgrade branch, which runs with this latest Sockethub (simply `npm install && npm start` with local Sockethub running):

https://github.com/67P/hyperchannel/pull/303

There's also a new error in the console about the client not being able to create some object, but it seemed to have no effect on functionality in my smoke testing.

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
Merge branch 'master' into fix/xmpp-determin-actor-type

silverbucket committed:
fix(platform-xmpp): track joined rooms to correctly type presence actor (issue #679)

Replace the brittle service-discovery approach with proactive room tracking:

- join() registers the room bare JID in this.__knownRooms
- leave() removes it
- presence() checks __knownRooms synchronously — no network, no heuristics,
  no caching bugs
- Remove determineActorType(), jidTypeCache, DISCOVERY_TIMEOUT entirely

This correctly types kosmos@kosmos.chat (and any JID without 'conference.'
in the domain) as "room" because the information comes from the join call,
not from pattern-matching on the JID string.

Tests: unit tests cover registration/deregistration, resource stripping,
multi-room tracking, and correct AS object structure. Integration test
verifies that a presence update from a joined room has actor.type === "room".

silverbucket committed:
fix(platform-xmpp): use real MUC room in presence actor-type integration test

- Use config.prosody.room (testroom@conference.prosody) instead of the
  hardcoded non-MUC JID test@prosody so the MUC server actually sends
  back presence stanzas that the test can observe.
- Collapse multi-line platformIdFromContext() call to satisfy Biome
  formatter.
- Fix join() to build the MUC 'to' attribute from roomJid (bare JID)
  rather than job.target.id, avoiding double-resource JIDs when the
  target carries a resource portion.
- Update the incoming-message-queue assertion to allow presence pushes
  from the real MUC join alongside the existing error-message allowance.

silverbucket commented on integration/browser/basic.integration.js:
```diff
@@ -286,6 +287,44 @@ describe(`Sockethub Basic Integration Tests at ${config.sockethub.url}`, () => {
+                        expect(roomPresences[0].actor.type).to.equal("room");
+                        expect(
+                            platformIdFromContext(
+                                roomPresences[0]["@context"],
+                            ),
+                        ).to.equal("xmpp");
```
Fixed in the latest commit — the multi-line call has been collapsed to a single line.

silverbucket commented on packages/platform-xmpp/src/incoming-handlers.js:
```diff
@@ -60,6 +61,95 @@ function getPresence(stanza) {
+
+            // Default to person if no room indicators found
+            this.jidTypeCache.set(bareJid, "person");
+            return "person";
```
This entire code path no longer exists. The `determineActorType`/`jidTypeCache`/disco#info approach has been replaced by synchronous lookup against the `__knownRooms` Set that `join()` and `leave()` maintain.

silverbucket commented on packages/platform-xmpp/src/incoming-handlers.js:
```diff
@@ -60,6 +61,95 @@ function getPresence(stanza) {
+                    : "error";
+            this.session.debug(
+                `Service discovery ${errorType} for ${bareJid}, falling back to heuristic: ${error.message}`,
+            );
```
This code is gone — the reactive disco#info approach (and the `this.session.debug` bug) was replaced entirely.

silverbucket commented on packages/platform-xmpp/src/incoming-handlers.js:
```diff
@@ -60,6 +61,95 @@ function getPresence(stanza) {
+                    ? "room"
+                    : "person";
+            this.jidTypeCache.set(bareJid, type);
+            return type;
```
This code is gone — the `jidTypeCache` was replaced by the `__knownRooms` Set in XMPP, which is authoritative (populated by `join()`, cleared by `leave()`) and never stale.

silverbucket committed:
chore: lint

silverbucket committed:
fix(platform-xmpp): defer __knownRooms mutations until send succeeds, fix leave() JID, clear on cleanup

- Move __knownRooms.add() into the join() success callback so a failed
  send doesn't permanently mark a room as joined.
- Move __knownRooms.delete() into the leave() success callback and
  restore the entry on failure, keeping the set consistent.
- Propagate leave() send errors to done() instead of silently swallowing them.
- Replace job.target?.id with roomJid in leave() to avoid double-resource JIDs.
- Clear __knownRooms in cleanup() so stale room state doesn't persist
  across reconnections.

silverbucket committed:
test(platform-xmpp): cover room tracking regressions

raucao reviewed (approved):
This finally works reliably! 👍
